pkg/reconciler/dns/health.go

```
@@ -0,0 +1,191 @@
+	}
+
+	if config == nil {
+		return c.reconcileHealthCheckDeletion(ctx, dnsRecord)
+	}
+
+	if err := validateHealthChecksConfig(config); err != nil {
+		return err
+	}
+
+	return c.reconcileHealthCheck(ctx, config, dnsRecord)
+}
+
+func (c *Controller) reconcileHealthCheck(ctx context.Context, config *healthChecksConfig, dnsRecord *v1.DNSRecord) error {
+	healthCheck := c.dnsProvider.HealthCheckReconciler()
+
+	for _, dnsEndpoint := range dnsRecord.Spec.Endpoints {
+		ok := false
+		if _, ok = dnsEndpoint.GetAddress(); !ok {
+			klog.Infof("Skipping health check creation for %s/%s. No address set", dnsRecord.Name, dnsEndpoint.DNSName)
+			continue
+		}
+
+		spec := dns.HealthCheckSpec{
+			Id:               idForEndpoint(dnsRecord, dnsEndpoint),
+			Name:             fmt.Sprintf("%s-%s", dnsEndpoint.DNSName, dnsEndpoint.SetIdentifier),
+			Path:             config.Endpoint,
+			Port:             config.Port,
+			Protocol:         config.Protocol,
+			FailureThreshold: config.FailureThreshold,
+		}
+
+		klog.Infof("Reconciling health check for endpoint %s@%s", dnsEndpoint.DNSName, dnsEndpoint.SetIdentifier)
+
+		err := healthCheck.Reconcile(ctx, spec, dnsEndpoint)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *Controller) reconcileHealthCheckDeletion(ctx context.Context, dnsRecord *v1.DNSRecord) error {
+	reconciler := c.dnsProvider.HealthCheckReconciler()
+
+	for _, zone := range dnsRecord.Status.Zones {
+		for _, endpoint := range zone.Endpoints {
+			if err := reconciler.Delete(ctx, endpoint); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// idForEndpoint returns a unique identifier for an endpoint
+func idForEndpoint(dnsRecord *v1.DNSRecord, endpoint *v1.Endpoint) string {
+	hash := md5.New()
+	if _, err := io.WriteString(hash, fmt.Sprintf("%s/%s@%s", dnsRecord.Name, endpoint.SetIdentifier, endpoint.DNSName)); err != nil {
+		klog.Errorf("Unexpected error creating ID for endpoint %s", endpoint.SetIdentifier)
```

Do we want to return an error here?

---

Yes, I updated it to return the error